### PR TITLE
[CLC-245]: Add Command to List Project Templates

### PR DESCRIPTION
**Pull request by kutluhanmetin**

### CODE REVIEW

File: base/commands/project/project_list_templates.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	}
+	for _, t := range ts {
+		templates = append(templates, Template{Name: t, Source: "local"})
+	}
+	return templates, nil
+}
+
+func fetchTemplates() ([]Template, error) {
+	var templates []Template
+	resp, err := http.Get(fetchURL())
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:125]: Should the results be paged? There maybe hundreds of repos in the org.

File: base/commands/project/project_list_templates.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	if err != nil {
+		return []Template{}, err
+	}
+	respData, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return []Template{}, err
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:131]:
```suggestion
		return nil, err
```

File: base/commands/project/project_list_templates.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		templates = append(templates, Template{Name: t, Source: "local"})
+	}
+	return templates, nil
+}
+
+func fetchTemplates() ([]Template, error) {
+	var templates []Template
+	resp, err := http.Get(fetchURL())
+	if err != nil {
+		return []Template{}, err
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:127]:
```suggestion
		return nil, err
```

File: base/commands/project/project_list_templates.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+	}
+	respData, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return []Template{}, err
+	}
+	var data []map[string]any
+	err = json.Unmarshal(respData, &data)
+	if err != nil {
+		return []Template{}, err
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:136]:
```suggestion
		return nil, err
```

File: base/commands/project/project_list_templates.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+	}
+	for _, d := range data {
+		var tName string
+		var ok bool
+		if tName, ok = d["full_name"].(string); !ok {
+			return []Template{}, errors.New("error fetching repositories in the organization")
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:142]:
```suggestion
			return nil, errors.New("error fetching repositories in the organization")
```

File: base/commands/project/project_list_templates.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+		name := sName[1]
+		templates = append(templates, Template{Name: name, Source: source})
+	}
+	return templates, nil
+}
+
+func updateNextFetchTime(sa *store.StoreAccessor) error {
+	_, err := sa.WithLock(func(s *store.Store) (any, error) {
+		return nil, s.SetEntry(bytes(nextFetchTimeKey),
+			bytes(strconv.FormatInt(time.Now().Add(cacheRefreshInterval).Unix(), 10)))
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:155]: Would it make sense to move the value to a variable?

File: base/commands/project/project_list_templates.go [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+	})
+	return err
+}
+
+func fetchURL() string {
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:160]:
```suggestion
func makeRepositoriesURL() string {
```

File: base/commands/project/project_list_templates.go [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+	s := strings.TrimPrefix(templateOrgURL(), "https://github.com/")
+	ss := strings.ReplaceAll(s, "/", "")
+	return fmt.Sprintf("https://api.github.com/users/%s/repos", ss)
+}
+
+func isFetch(s *store.StoreAccessor) (bool, error) {
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:166]:
```suggestion
func canFetchTemplateList(s *store.StoreAccessor) (bool, error) {
```

File: base/commands/project/project_list_templates.go [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+	stop()
+	return ec.AddOutputRows(ctx, rows...)
+}
+
+func listTemplates(logger log.Logger, isLocal bool, isForce bool) ([]Template, error) {
+	sa := store.NewStoreAccessor(paths.Join(paths.Home(), storeFolder), logger)
+	if isLocal {
+		return listLocalTemplates()
+	}
+	if fetch, err := isFetch(sa); err != nil {
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:94]: Can we continue and list the templates if `isFetch` returns an error?
Logging the error in DEBUG would be good.

File: base/commands/project/project_list_templates.go [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+		return []Template{}, err
+	} else if fetch || isForce {
+		ts, err := fetchTemplates()
+		if err != nil {
+			return nil, err
+		}
+		err = updateCache(sa, ts)
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:101]: Let's log the error as DEBUG and continue listing the templates.

File: base/commands/project/project_list_templates.go [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+func updateCache(sa *store.StoreAccessor, templates []Template) error {
+	b, err := json.Marshal(templates)
+	if err != nil {
+		return err
+	}
+	_, err = sa.WithLock(func(s *store.Store) (any, error) {
+		err = s.DeleteEntriesWithPrefix(templatesKey)
+		if err != nil {
+			return nil, err
+		}
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:197]: Do we have to delete the template key?
Wouldn't setting the new value overwrite the older one?

File: base/commands/project/project_list_templates.go [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+		err = s.SetEntry(bytes(templatesKey), b)
+		if err != nil {
+			return nil, err
+		}
+		return nil, nil
+	})
+	if err != nil {
+		return err
+	}
+	if err = updateNextFetchTime(sa); err != nil {
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:207]: Would it make sense to update the fetch time in the .SetEntry block above?

File: base/commands/project/project_list_templates.go [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+	return nil
+}
+
+func listFromCache(sa *store.StoreAccessor) ([]Template, error) {
+	var templates []Template
+	b, err := sa.WithLock(func(s *store.Store) (any, error) {
+		return s.GetEntry(bytes(templatesKey))
+	})
+	if err != nil {
+		return []Template{}, err
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:219]:
```suggestion
		return nil, err
```

File: base/commands/project/project_list_templates.go [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+	}
+	err = json.Unmarshal(b.([]byte), &templates)
+	if err != nil {
+		return []Template{}, err
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:223]:
```suggestion
		return nil, err
```

File: base/commands/project/project_list_templates.go [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+	}
+	return templates, nil
+}
+
+func bytes(s string) []byte {
+	return []byte(s)
+}
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:230]: `[]byte(s)` is quite idiomatic, so I think `bytes` function is not necessary.

**Review by yuce — CHANGES_REQUESTED** [review]
Some ideas...

File: base/commands/project/project_list_it_test.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+		sPath := filepath.Join(tcx.HomePath(), storeFolder)
+		sa := store.NewStoreAccessor(sPath, log.NopLogger{})
+		check.MustValue(sa.WithLock(func(s *store.Store) (any, error) {
+			err := s.SetEntry(bytes(nextFetchTimeKey),
+				bytes(strconv.FormatInt(time.Now().Add(cacheRefreshInterval).Unix(), 10)))
+			return nil, err
+		}))
+		check.MustValue(sa.WithLock(func(s *store.Store) (any, error) {
+			b, err := json.Marshal([]Template{{Name: "test_template"}})
+			check.Must(err)
```
[the remark this answers is not in the document]
Comment (kutluhanmetin) [reply base/commands/project/project_list_it_test.go:44]: I did it on purpose that I need `b` on line 45: `err = s.SetEntry(bytes(templatesKey), b)`

File: base/commands/project/project_list_templates.go (hunk 8 above)
Comment (kutluhanmetin) [reply base/commands/project/project_list_templates.go:166]: I think we should not use the word `can` that it is not related to ability. We are checking if we need to fetch the template list from remote or not.

File: base/commands/project/project_list_templates.go (hunk 1 above)
Comment (kutluhanmetin) [reply base/commands/project/project_list_templates.go:125]: Let me check if the API provides paging 👍

Comment (kutluhanmetin) [reply base/commands/project/project_list_templates.go]: Yes, it is possible via API: https://api.github.com/users/hazelcast-templates/repos?per_page=10&page=1, but how should be our user experience to switch to the next page in CLC?

File: base/commands/project/project_list_templates.go [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+
+const (
+	flagLocal            = "local"
+	flagForce            = "force"
+	nextFetchTimeKey     = "project.templates.nextFetchTime"
+	templatesKey         = "project.templates"
+	cacheRefreshInterval = 10 * time.Minute
+)
+
+var storeFolder = filepath.Join("caches", "templates")
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:36]: Could you add the following function in `paths.go`
```

func Caches() string {
	return filepath.Join(Home(), "caches")
}
```
I think with that change we can remove `storeFolder`.

File: base/commands/project/project_list_templates.go [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+	cc.AddBoolFlag(flagLocal, "", false, false, "list the templates which exist on local environment")
+	help := "Lists templates that can be used while creating projects."
+	cc.SetCommandHelp(help, help)
+	return nil
+}
+
+func (lc ListCmd) Exec(ctx context.Context, ec plug.ExecContext) error {
+	isLocal := ec.Props().GetBool(flagLocal)
+	isForce := ec.Props().GetBool(flagForce)
+	if isLocal && isForce {
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:56]:
```suggestion
	local := ec.Props().GetBool(flagLocal)
	refresh := ec.Props().GetBool(flagForce)
	if local && refresh {
```

File: base/commands/project/project_list_templates.go [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+		fetch = true
+	}
+	if fetch || isForce {
+		ts, err := fetchTemplates()
+		if err != nil {
+			return nil, err
+		}
+		err = updateCache(sa, ts)
+		if err != nil {
+			logger.Debugf("Updating templates cache: %w", err)
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:108]: Just to communicate that this is an error:
```suggestion
			logger.Debugf("Error: updating templates cache: %w", err)
```

File: base/commands/project/project_list_templates.go (hunk 8 above)
Comment (yuce) [reply base/commands/project/project_list_templates.go:171]: How about `shouldFetch` ?

File: base/commands/project/project_list_templates.go [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+
+func listTemplates(logger log.Logger, isLocal bool, isForce bool) ([]Template, error) {
+	sa := store.NewStoreAccessor(paths.Join(paths.Home(), storeFolder), logger)
+	if isLocal {
+		return listLocalTemplates()
+	}
+	var fetch bool
+	var err error
+	if fetch, err = isFetch(sa); err != nil {
+		logger.Debugf("Decide to fetch templates: %w", err)
```
Comment (yuce) [inline base/commands/project/project_list_templates.go:97]:
```suggestion
		logger.Debugf("Error: checking template list expiry: %w", err)
```

**Review by yuce — CHANGES_REQUESTED** [review]
A few more ideas...

**Review by yuce — APPROVED** [review]
Looks good!

**Pull request merged**